dqn.py

```diff
@@ -0,0 +1,202 @@
+        done = torch.FloatTensor(done).to(self.device).reshape(batch_size)
+
+        q_value = self.policy_network(state).gather(1, action).squeeze(1)
+        next_q_value = self.target_network(next_state).max(1)[0]
+        expected_q_value = reward + self.gamma * next_q_value * (1 - done)
+        loss = self.criterion(q_value, expected_q_value.detach())
+
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+        self.step += 1
+        if self.step % self.target_update == 0:
+            self.target_network.load_state_dict(self.policy_network.state_dict())
+            self.step = 0
+
+    def update_target(self):
+        self.target_network.load_state_dict(self.policy_network.state_dict())
+
+    def get_action(self, state, epsilon):
+        if epsilon != 0 and np.random.rand() <= epsilon:
+            return np.random.randint(self.action_size)
+        else:
+            state = torch.FloatTensor(state).to(self.device)
+            q_value = self.target_network(state)
+            return q_value.max(1)[1].detach().cpu().numpy() if len(
+                q_value.shape) > 1 else q_value.max(0)[1].detach().cpu().numpy()
+
+    def get_policy_f(self):
+        from functools import partial
+        return partial(self.get_action, epsilon=0.0)
+
+    def get_state_value(self, state):
+        state = torch.FloatTensor(state).to(self.device)
+        q_value = self.target_network(state)
+        # get the max value of the state and flatten it
+        return q_value.max(1)[0].detach().cpu().numpy()
+
+    def save_model(self, path):
+        torch.save(self.target_network.state_dict(), path)
+
+    def load_model(self, path):
+        # check path exists
+        if os.path.isfile(path):
+            print("Loading model from {}".format(path))
+            self.target_network.load_state_dict(torch.load(path))
+        else:
+            print("No model found at {}".format(path))
+
+
+def train(env, agent, num_episodes, batch_size, epsilon, epsilon_decay, epsilon_min, render=False):
+    # 2d np array to store the scores
+    scores = np.zeros([num_episodes, env.max_num_steps], dtype=np.float32)
+    score = np.zeros(env.max_num_steps, dtype=np.float32)
+    # make a progress bar to show the training progress
+
+    for e in range(num_episodes):
+        state = env.reset()
+        step = 0
+        while True:
+            if render:
+                env.render()
+            action = agent.get_action(state, epsilon)
+            next_state, reward, done = env.step(action)
+            agent.buffer.add(state, action, reward, next_state, done)
+            state = next_state
+            score[step] = reward
+            step += 1
+            if len(agent.buffer) >= batch_size:
+                agent.update(batch_size)
+            if done:
+                break
+        epsilon = max(epsilon * epsilon_decay, epsilon_min)
```

Comment:
You could do the epsilon decay within your get action function. It would make more sense to do it there. (that may be personal preference however)

Reply:
You are right. It would be better for parameter management